tests/simulated_device_and_reader.rs

```diff
@@ -0,0 +1,131 @@
+        .context("failed to initialize device")?;
+
+    session
+        .qr_engagement()
+        .context("could not generate qr engagement")
+}
+
+/// Parse the mDL encoded string into a [Documents] object.
+fn parse_mdl() -> Result<NonEmptyMap<DocType, Document>> {
+    let mdl_encoded = include_str!("data/stringified-mdl.txt");
+    let mdl = Document::parse(mdl_encoded.to_string()).context("could not parse mDL")?;
+    let docs = Documents::new(DOC_TYPE.to_string(), mdl);
+    Ok(docs)
+}
+/// Establishes the reader session from the given QR code and create request for needed elements.
+fn establish_reader_session(qr: String) -> Result<(reader::SessionManager, Vec<u8>)> {
+    let requested_elements = Namespaces::new(
+        NAMESPACE.into(),
+        DataElements::new(AGE_OVER_21_ELEMENT.to_string(), false),
+    );
+    let (reader_sm, session_request, _ble_ident) =
+        reader::SessionManager::establish_session(qr, requested_elements)
+            .context("failed to establish reader session")?;
+    Ok((reader_sm, session_request))
+}
+
+/// The Device handles the request from the reader and advances the state.
+fn handle_request(
+    state: SessionManagerEngaged,
+    request: Vec<u8>,
+) -> Result<(device::SessionManager, RequestedItems)> {
+    let (session_manager, items_requests) = {
+        let session_establishment: definitions::SessionEstablishment =
+            serde_cbor::from_slice(&request).context("could not deserialize request")?;
+        state
+            .process_session_establishment(session_establishment)
+            .context("could not process process session establishment")?
+    };
+    if session_manager.get_next_signature_payload().is_some() {
+        anyhow::bail!("there were errors processing request");
+    }
+    Ok((session_manager, items_requests))
+}
+
+// Prepare response with required elements.
+fn create_response(
+    mut session_manager: device::SessionManager,
+    requested_items: RequestedItems,
+    key: &p256::ecdsa::SigningKey,
+) -> Result<Vec<u8>> {
+    let permitted_items = [(
+        DOC_TYPE.to_string(),
+        [(NAMESPACE.to_string(), vec![AGE_OVER_21_ELEMENT.to_string()])]
+            .into_iter()
+            .collect(),
+    )]
+    .into_iter()
+    .collect();
+    session_manager.prepare_response(&requested_items, permitted_items);
+    let (_, sign_payload) = session_manager.get_next_signature_payload().unwrap();
+    let signature: p256::ecdsa::Signature = key.sign(sign_payload);
+    session_manager
+        .submit_next_signature(signature.to_vec())
+        .context("failed to submit signature")?;
+    session_manager
+        .retrieve_response()
+        .ok_or(anyhow!("cannot prepare response"))
+}
+
+/// Reader Processing mDL data.
+fn reader_handle_device_response(
+    reader_sm: &mut reader::SessionManager,
+    response: Vec<u8>,
+) -> Result<()> {
+    let res = reader_sm.handle_response(&response)?;
+    println!("{:?}", res);
+    Ok(())
+}
```

Comment:
I think it's confusing that these functions are just duplicated from `common.rs` why not just import them?

Reply:
Should I just keep these 2? `simulated_device_and_reader` and `simulated_device_and_reader_state`?

Reply:
Moved methods to `Device` and `Reader` and now the tests are less code, maybe we could keep them as we include them in docs examples.

Reply:
Do you think it is valuable to keep both? The one using `Arc` and `Mutex` doesn't add much value IMO because in any real application you wouldn't need to do this.

Reply:
I removed `on_simulated_device` and `on_simulated_reader` and we left with only 2.